tools/run-fbp-tests

```diff
@@ -39,11 +39,106 @@ from io import open
 
 import argparse
 import difflib
+import logging
 import os
 import re
 import subprocess
 import sys
 
+DEFAULT_VALGRIND_OPTS = (
+    "--show-reachable=no",
+    "--quiet",
+    "--error-exitcode=1",
+    "--leak-check=full",
+)
+
+class FbpTest:
+    SKIP_NONE = 0
+    SKIP_SIMPLE = 1
+    SKIP_VALGRIND = 2
+    SKIP_COMPILE = 3
+
+    SUPPORTED_SKIP = {
+        "TEST-SKIP": SKIP_SIMPLE,
+        "TEST-SKIP-VALGRIND": SKIP_VALGRIND,
+        "TEST-SKIP-COMPILE": SKIP_COMPILE,
+    }
+
+    def __init__(self, test_file):
+        self.test_file = test_file
+        self.skip_msg = None
+        self.skip = self.SKIP_NONE
+        self.output = False
+        self.output_regex = False
+        self.output_spec = ""
+        self.exit_code = 0
+
+        self._parse_test()
+
+    def _parse_skip(self, cmd, splitted):
+        self.skip = self.SUPPORTED_SKIP.get(cmd, self.SKIP_NONE)
+        if not self.skip:
+            return
+
+        if len(splitted) > 2:
+            self.skip_msg = splitted[2]
+
+    def _parse_command(self, line):
+        splitted = line.split(None, 2)
+        cmd = splitted[1]
+
+        if cmd == "TEST-OUTPUT-REGEX":
+            self.output_regex = True
+        elif cmd == "TEST-OUTPUT":
+            self.output = True
+        elif cmd == "TEST-EXPECTS-ERROR":
+            self.exit_code = 1
+        elif cmd.startswith("TEST-SKIP"):
+            self._parse_skip(cmd, splitted)
+        else:
+            logging.info("Ignoring unknown test command '%s' in test %s" %
+                         (cmd, self.test_file))
+
+    def _parse_test(self):
+        for line in open(self.test_file, encoding="UTF-8"):
+            if line.startswith("## "):
+                self._parse_command(line)
+            elif (self.output or self.output_regex) and line.startswith("#"):
+                self.output_spec += line[2:] or "\n"
+
+    def should_skip(self, skips):
+        if self.skip == self.SKIP_NONE:
+            return False
+        elif self.skip in skips:
+            return True
+        return False
+
+    def show_result_status(self, out, code):
+        if self.output and self.output_spec != out:
+            orig = out.splitlines(True)
+            dest = self.output_spec.splitlines(True)
+            logging.warning("FAILED running %s: wrong output" % os.path.basename(self.test_file))
+            for line in difflib.unified_diff(orig, dest):
+                sys.stdout.write(line)
+            return "failed"
```

Comment:
I use a dict to accumulate the status, so that's the why I returning the key part but we can change it, I just didn't want to have 3 if - else blocks to define the var to use and so on, latter on I do iterate over this dict items to print the status, just a matter of keeping the code a bit simpler.